pkg/registry/blobs.go

```diff
@@ -220,21 +352,43 @@ func (b *blobs) handle(resp http.ResponseWriter, req *http.Request) *regError {
 
 		b.lock.Lock()
 		defer b.lock.Unlock()
-		l := bytes.NewBuffer(b.uploads[target])
-		io.Copy(l, req.Body)
-		rd := sha256.Sum256(l.Bytes())
-		d := "sha256:" + hex.EncodeToString(rd[:])
-		if d != digest {
+
+		h, err := v1.NewHash(digest)
+		if err != nil {
 			return &regError{
 				Status:  http.StatusBadRequest,
-				Code:    "DIGEST_INVALID",
-				Message: "digest does not match contents",
+				Code:    "NAME_INVALID",
+				Message: "invalid digest",
+			}
+		}
+
+		defer req.Body.Close()
+		in := ioutil.NopCloser(io.MultiReader(bytes.NewBuffer(b.uploads[target]), req.Body))
+
+		size := int64(verify.SizeUnknown)
+		if req.ContentLength > 0 {
+			size = int64(len(b.uploads[target])) + req.ContentLength
+		}
+
+		vrc, err := verify.ReadCloser(in, size, h)
+		if err != nil {
+			return errTODO(err.Error())
+		}
+		defer vrc.Close()
```

Comment:
Don't you have to check the error here for the verify.ReadCloser to be meaningful at all?

Reply:
`verify.ReadCloser` verifies when `inner.Read` hits and `EOF`, not on close.

Reply:
At one point this is exactly how Docker checked digests and there was a bug that turned out to be an unchecked error on (drumroll) `io.Copy` 😬 

I haven't looked carefully to know if this has that problem, but there are lots of unchecked `io.Copy` calls in this file.

Reply:
Checking those errors exposed a bug in `verify.ReadCloser` when the size isn't known ahead of time. Fixed, with test that fails without the change. 🐛 🔨